toolchain/lower/file_context.cpp

```diff
@@ -121,6 +137,219 @@ auto FileContext::Run() -> std::unique_ptr<llvm::Module> {
+                                     .index] !=
+           equivalent_specifics_[specific_to_update.index]) {
+      stack.push_back(specific_to_update);
+      specific_to_update = equivalent_specifics_[specific_to_update.index];
+    }
+    for (auto specific : std::ranges::reverse_view(stack)) {
+      equivalent_specifics_[specific.index] =
+          equivalent_specifics_[equivalent_specifics_[specific.index].index];
+    }
+  };
+
+  update_equivalent_specific(specific1);
+  update_equivalent_specific(specific2);
+
+  return equivalent_specifics_[specific1.index] ==
+         equivalent_specifics_[specific2.index];
+}
+
+auto FileContext::DeleteFunctionSpecific(SemIR::SpecificId to_replace) -> void {
+  specific_functions_[to_replace.index]->eraseFromParent();
+  specific_functions_[to_replace.index] =
+      specific_functions_[equivalent_specifics_[to_replace.index].index];
+}
+
+auto FileContext::AreFunctionTypesEquivalent(SemIR::SpecificId specific1,
+                                             SemIR::SpecificId specific2)
+    -> bool {
+  CARBON_CHECK(specific1.has_value() && specific2.has_value());
+  return lowered_specifics_type_fingerprint_[specific1.index] ==
+         lowered_specifics_type_fingerprint_[specific2.index];
+}
+
+auto FileContext::AreFunctionBodiesEquivalent(
+    SemIR::SpecificId specific1, SemIR::SpecificId specific2,
+    Set<std::pair<SemIR::SpecificId, SemIR::SpecificId>>&
+        visited_equivalent_specifics,
+    Set<std::pair<SemIR::SpecificId, SemIR::SpecificId>>&
+        visited_equivalent_specifics_flipped) -> bool {
+  llvm::SmallVector<std::pair<SemIR::SpecificId, SemIR::SpecificId>> worklist;
+  worklist.push_back({specific1, specific2});
+
+  while (!worklist.empty()) {
+    auto [specific_id1, specific_id2] = worklist.back();
+    worklist.pop_back();
+    std::pair<SemIR::SpecificId, SemIR::SpecificId> outer_pair = {specific_id1,
+                                                                  specific_id2};
+
+    auto state1 = lowered_specific_fingerprint_[specific_id1.index];
+    auto state2 = lowered_specific_fingerprint_[specific_id2.index];
+    if (state1.function_common_fingerprint !=
+        state2.function_common_fingerprint) {
+      auto insert_result = non_equivalent_specifics_.Insert(outer_pair);
+      // Save a lookup; if pair was previously inserted, so was the flipped
+      if (insert_result.is_inserted()) {
+        non_equivalent_specifics_.Insert(
+            std::make_pair(specific_id2, specific_id1));
+      }
+      return false;
+    }
+    if (state1.function_specific_fingerprint ==
+        state2.function_specific_fingerprint) {
+      return true;
+    }
+
+    // A size difference should have been detected by the common fingerprint.
+    CARBON_CHECK(state1.calls.size() == state2.calls.size(),
+                 "Number of specific calls expected to be the same.");
+
+    for (unsigned i = 0; i < state1.calls.size(); ++i) {
```

Comment:
Note [style discourages](https://google.github.io/styleguide/cppguide.html#Integer_Types) use of unsigned. Maybe use `llvm::zip` here, something like:

```suggestion
    for (auto [state1_call, state2_call] = llvm::zip(state1.calls, state2.calls)) {
```